### PR TITLE
Snake case

### PR DESCRIPTION
**Pull request by robhuzzey**
#### What does this PR do?
Switches the filenames generated to snake case as this is meant to be the standard.

#### What unit or integration tests does this PR have?
All the usual unit & system tests

#### What selenium tests does this PR have?
N/A

#### How should a developer review this?
Ensure unit tests work & coding matches expected standard.

#### How should this be manually tested?
Run the unit tests

#### Any background context you want to provide?
I was using this generator during project lounge but was told by Mike Carter that we should use snakecase for filenames... this is my attempt at a fix.

#### What are the relevant tickets?
N/A

#### Screenshots (if appropriate)
N/A

#### What gif best describes how you feel about this work?
Fun times

#### Quality checklist (for PR author to complete BEFORE code review):
- [x] I've checked there is appropriate unit test coverage.
- [x] I've checked there is appropriate selenium test coverage.
- [x] I've included regression tests for any bug fixes.
- [x] I've updated documentation with any changes to procedures.
- [x] I've tested this on mobile Android, Safari, Chrome for any visual changes.
- [x] I've tested this on desktop IE9/10/11, Firefox, Chrome for any visual changes.
- [x] I've checked any significant visual changes with a member of the UX/UI pod.
- [x] I've checked this work against the requirements of the Jira.
- [x] I've added and updated translations for all supported languages
- [x] I've informed those who need to know of any unforeseen impact of this work (database migrations, deployment procedure changes).
- [x] I am ready for this to be code reviewed, merged and tested.
- [x] I've thought about writing a blog post about this

#### IP or Developer Review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] +1 from me!

#### Software Engineer or Developer review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I've checked this work against the requirements of the Jira.
- [x] +1 from me!

#### Software Engineer or project guru review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I don’t believe this work introduces unnecessary technical debt.
- [x] +1 from me!

### CODE REVIEW

**Comment by LouisPryer** [comment]
Taking IP/Dev

**Comment by LouisPryer** [comment]
:+1:

File: tests/generators/base/webAppViewBaseGeneratorTest.js [hunk 1]
```diff
@@ -47,4 +47,102 @@ describe('ui-component:base generator', function () {
...
   });
+  describe( '_toSnakeCase()', function() {
+    describe('no input', function() {
+      it('should return an empty string', function() {
+        assert.strictEqual('', app._toSnakeCase())
+      });
+    });
+    describe('mixed casing', function() {
+      it('should return a lowercase string', function() {
+        assert.strictEqual('foobar', app._toSnakeCase('fOoBaR'))
```
Comment (lukehansell-hx) [inline tests/generators/base/webAppViewBaseGeneratorTest.js]: should we expect this to return "foobar" or "f_oo_ba_r"? I'd argue the latter.

File: tests/generators/base/webAppViewBaseGeneratorTest.js [hunk 2]
```diff
@@ -47,4 +47,102 @@ describe('ui-component:base generator', function () {
...
+  });
+  describe( '_generateComponentName()', function() {
+    describe('no input', function() {
+      it('should return an empty string', function() {
+        assert.strictEqual('', app._generateComponentName())
+      });
+    });
+    describe('mixed casing', function() {
+      it('should return a string with the correct camel casing & \"Component\" appended', function() {
+        assert.strictEqual('FoobarComponent', app._generateComponentName('fOoBaR'))
```
Comment (lukehansell-hx) [inline tests/generators/base/webAppViewBaseGeneratorTest.js]: same as above really, I would expect "FOoBaRComponent" as "fOoBaR" appears to already be camel cased.

File: tests/generators/base/webAppViewBaseGeneratorTest.js [hunk 3]
```diff
@@ -47,4 +47,102 @@ describe('ui-component:base generator', function () {
...
+      });
+    });
+    describe('hyphens in the string', function() {
+      it('should return a string with the correct camel casing & \"Component\" appended', function() {
+        assert.strictEqual('FooBarBazComponent', app._generateComponentName('foo-bar--baz'))
+      });
+    });
+    describe('spaces in the string', function() {
```
Comment (lukehansell-hx) [inline tests/generators/base/webAppViewBaseGeneratorTest.js]: do we require this test? Spaces are not valid as input according to the _validateName method.

File: generators/base/index.js [hunk 4]
```diff
@@ -44,6 +44,30 @@ module.exports = generators.Base.extend({
   },
   
   _capitalize: function(input){
-    return input.charAt(0).toUpperCase() + input.slice(1)
+    return input.charAt(0).toUpperCase() + input.slice(1);
+  },
+
+  _toSnakeCase: function(input) {
```
Comment (lukehansell-hx) [inline generators/base/index.js]: I would argue that this method should snake case camel cased words. e.g. "fOoBa-r" should return "f_oo_ba_r".

**Comment by michaelcarter** [comment]
Shotgun SE/Dev

File: tests/generators/base/webAppViewBaseGeneratorTest.js (hunk 1 above)
Comment (robhuzzey) [inline tests/generators/base/webAppViewBaseGeneratorTest.js]: Would we want to convert camelcasing to snakecasing as that's what you are suggesting?

Comment (lukehansell-hx) [inline tests/generators/base/webAppViewBaseGeneratorTest.js]: That makes most sense to me instead of just lowercasing everything.

**Comment by lukehansell-hx** [comment]
taking SE/Guru

File: generators/base/index.js (hunk 4 above)
Comment (lukehansell-hx) [inline generators/base/index.js]: alternatively _validateName could remove the ability to enter upper case characters and dashes with a note requesting a snake cased name?

File: generators/base/index.js [hunk 5]
```diff
@@ -44,6 +44,30 @@ module.exports = generators.Base.extend({
...
+  },
+
+  _generateFileName: function(input, ext) {
+    if(!input) return '';
+    return [this._toSnakeCase(input), (ext || 'jsx')].join('.')
+  },
+
+  _generateName: function(input) {
+    if(!input) return '';
+    return this._toSnakeCase(input).replace(/_/g, '-');
```
Comment (michaelcarter) [inline generators/base/index.js]: See the [underscore string](http://epeli.github.io/underscore.string/#api) methods for this stuff: `titleize`, `camelize`, `classify`, `underscored`, `dasherize` even `humanize`. It won't save you too much code but they might hit a few edge cases you may have missed here and are definitely more readable.

**Comment by michaelcarter** [comment]
The changes look sensible for the most part, I agree that `fooBar` to snake case should be `foo_bar` rather than `foobar` though.

I think we'd be better off switching to the underscore.string library for some of this string manipulation work.

**Comment by robhuzzey** [comment]
@michaelcarter @lukehansell-hx I've updated to use lodash string library (faster than underscore) and in addition restricted the input text to only a-z (lowercase) and underscores.

I have still left in the string manipulation just in case something changes in the future that affects the validation.

Please re-review. Thanks.

**Comment by michaelcarter** [comment]
nodding

+1

**Comment by lukehansell-hx** [comment]
looks good :+1:

**Pull request merged**